bee-protocol/src/workers/message/hasher.rs

```diff
@@ -0,0 +1,153 @@
+
+use std::{any::TypeId, convert::Infallible};
+
+pub(crate) struct HasherWorkerEvent {
+    pub(crate) from: Option<PeerId>,
+    pub(crate) message_packet: MessagePacket,
+    pub(crate) notifier: Option<Sender<Result<MessageId, MessageSubmitterError>>>,
+}
+
+pub(crate) struct HasherWorker {
+    pub(crate) tx: mpsc::UnboundedSender<HasherWorkerEvent>,
+}
+
+#[async_trait]
+impl<N: Node> Worker<N> for HasherWorker
+where
+    N::Backend: StorageBackend,
+{
+    type Config = ProtocolConfig;
+    type Error = Infallible;
+
+    fn dependencies() -> &'static [TypeId] {
+        vec![
+            TypeId::of::<ProcessorWorker>(),
+            TypeId::of::<MetricsWorker>(),
+            TypeId::of::<PeerManagerResWorker>(),
+        ]
+        .leak()
+    }
+
+    async fn start(node: &mut N, config: Self::Config) -> Result<Self, Self::Error> {
+        let (tx, rx) = mpsc::unbounded_channel();
+        let processor_worker = node.worker::<ProcessorWorker>().unwrap().tx.clone();
+        let metrics = node.resource::<NodeMetrics>();
+        let peer_manager = node.resource::<PeerManager>();
+
+        let minimum_pow_score = config.minimum_pow_score;
+
+        let mut cache = HashCache::new(config.workers.message_worker_cache);
+
+        node.spawn::<Self, _, _>(|shutdown| async move {
+            let mut receiver = ShutdownStream::new(shutdown, UnboundedReceiverStream::new(rx));
+            let mut hasher = UnrolledCurlP81::new();
+            let mut blake2b = Blake2b256::new();
+            let mut pow_input: TritBuf = TritBuf::with_capacity(243);
+
+            info!("Running.");
+
+            while let Some(HasherWorkerEvent {
+                from,
+                message_packet,
+                notifier,
+            }) = receiver.next().await
+            {
+                if !cache.insert(&message_packet.bytes) {
+                    // If the message was already received, we skip it and poll again.
+                    trace!("Message already received.");
+
+                    if let Some(notifier) = notifier {
+                        if let Err(e) =
+                            notifier.send(Err(MessageSubmitterError("message already received".to_string())))
+                        {
+                            error!("failed to send error: {:?}.", e);
+                        }
+                    }
+
+                    metrics.known_messages_inc();
+                    if let Some(peer_id) = from {
+                        if let Some(peer) = peer_manager.get(&peer_id).await {
+                            (*peer).0.metrics().known_messages_inc();
+                        }
+                    }
+                    continue;
+                }
+
+                // TODO const
+                // TODO check
+                // TODO see if there is something we can reuse from pow crate
+                let (head, tail) = message_packet.bytes.split_at(message_packet.bytes.len() - 8);
```

Comment:
what does `const` mean here?

I feel that we need some sort of `fn get_as_array<const N: usize,T>(slice: &[T]) -> Option<[T; N]>` and maybe something analogous to `split_at` too.

Good candidates for the `bee-common` libary :trollface: